traits/trait_list_object.py

```diff
@@ -53,460 +88,755 @@ def __init__(self, index=0, removed=None, added=None):
+    def __iadd__(self, other):
+        """ Implements the self += value operator.
+
+        Parameters
+        ----------
+        other : Any
+            The item to add to the list.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                Slice ranging from current_length to new_length.
+            added : list
+                The  new values that were added.
+            removed : list
+                The empty list [].
+
+        """
+        self.extend(other)
+        return self
+
+    def __imul__(self, count):
+        """ Implements the self *= value operator.
+
+        Parameters
+        ----------
+        count : int
+            Number of times to duplicate the list.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                Slice ranging from current_length to new_length.
+            added : list
+                The  new values that were added.
+            removed : list
+                The empty list [].
+
+        """
+        if count > 1:
+            self.extend(self * (count - 1))
+        elif count == 0:
+            self[:] = []
+        return self
+
+    # ------------------------------------------------------------------------
+    # Private interface
+    # ------------------------------------------------------------------------
+
+    def _get_removed(self, index):
+        """ Compute removed values given index. """
+        try:
+            return self[index]
+        except IndexError:
+            return []
+
+    def _normalize(self, index):
+        if isinstance(index, slice):
+            return self._normalize_slice(index)
+        else:
+            return self._normalize_index(index)
+
+    def _normalize_index(self, index):
+        """ Normalize integer index to range 0 to len (inclusive). """
+        index = operator.index(index)
+        if index < 0:
+            return max(0, len(self) + index)
+        else:
+            return min(len(self), index)
+
+    def _normalize_slice(self, index):
+        """ Normalize slice start and stop to range 0 to len (inclusive). """
+        if index.step is None or index.step > 0:
+            if index.start is not None:
+                start = self._normalize_index(index.start)
+            else:
+                start = 0
+            if index.stop is not None:
+                stop = self._normalize_index(index.stop)
             else:
-                index = -1
+                stop = len(self)
+        else:
+            if index.start is not None:
+                start = self._normalize_index(index.start)
+            else:
+                start = len(self)
+            if index.stop is not None:
+                stop = self._normalize_index(index.stop)
+            else:
+                stop = 0
+
+        return slice(start, stop, index.step)
+
+    def _should_notify(self, removed, added):
+        try:
+            if added == removed:
```

Comment:
Could we honour `comparison_mode` here?
https://github.com/enthought/traits/blob/44ca101bcfa2e5a3c31d5fab141e892fd767793d/traits/ctrait.py#L102-L109

For #942, I need to switch to comparison mode to compare values by identity in order to get firing notifications for any nested list inside `List(List(List(List(...))))` to work.

Reply:
Here is the test I am fighting with:
https://github.com/enthought/traits/blob/d167da4cf6725f7a84c5865a5e310da65e89671d/poc/experiments.py#L362-L391

Reply:
Actually, it seems that the existing `List` behaviour already does not honour the comparison mode:
```
from traits.api import HasTraits, Any, List, Float, Trait
from traits.constants import ComparisonMode


class Foo(HasTraits):

    a = List(Any(comparison_mode=ComparisonMode.equality))


class Bar(HasTraits):

    a = List(Any(comparison_mode=ComparisonMode.identity))


def fire():
    print("fire")


foo = Foo(a=[[]])
foo.on_trait_change(fire, "a_items")

foo.a[slice(0, 1)] = [[]]
# Nothing fires

bar = Bar(a=[[]])
bar.on_trait_change(fire, "a_items")

bar.a[slice(0, 1)] = [[]]
# Nothing fires
```


I guess we will have to rework this when we come to EEP3.
Resolving for now as this `TraitList` still needs to work with the existing listener framework.